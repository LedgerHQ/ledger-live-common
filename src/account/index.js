// @flow

export * from "./helpers";
export * from "./addAccounts";
export * from "./support";
export * from "./importAccounts";
export * from "./serialization";
export * from "./accountId";
export * from "./accountName";
export * from "./ordering";
export * from "./groupOperations";
export * from "./pending";
export * from "./formatters";
