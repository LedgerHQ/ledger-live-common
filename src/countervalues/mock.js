// @flow
import type { CounterValuesState, Histodays, PollAPIPair } from "./types";
import type { Currency } from "../types";
import { formatCounterValueDay } from ".";
import { delay } from "../promise";

type Pair = {
  from: Currency,
  to: Currency,
  exchange: string,
  dateFrom: Date,
  dateTo: Date,
  rate: Date => number
};

export const genDateRange = ({ dateFrom, dateTo, rate }: Pair): Histodays => {
  const histodays = {};
  for (let d = dateFrom.getTime(); d < dateTo; d += 24 * 60 * 60 * 1000) {
    const day = new Date(d);
    histodays[formatCounterValueDay(day)] = rate(new Date(d));
  }
  histodays.latest = rate(new Date());
  return histodays;
};

export const genStoreState = (pairs: Pair[]): CounterValuesState => {
  const state = { rates: {} };
  pairs.forEach(pair => {
    const { from, to, exchange } = pair;
    const a = (state.rates[to.ticker] = state.rates[to.ticker] || {});
    const b = (a[from.ticker] = a[from.ticker] || {});
    b[exchange] = genDateRange(pair);
  });
  return state;
};

export const baseMockBTCRates = {
  BCH: 0.06102,
  BTG: 0.003239,
  DASH: 0.02314614,
  DCR: 0.004706,
  DGB: 0.00000249,
  DOGE: 5.392e-7,
  ETH: 3.1983e-12,
  ETC: 1.203e-13,
  ZEN: 0.00135623,
  KMD: 0.0002006,
  LTC: 0.015603,
  PPC: 0.011236,
  PIVX: 0.0001792,
  QTUM: 0.00055208,
  XSN: 0.0000241,
  XST: 0.0021319999999999998,
  STRAT: 0.0002083,
  VTC: 0.00009569,
  VIA: 0.0001065,
  XRP: 0.006208,
  ZEC: 0.0133534
};

export const getDailyRatesImplementation = async (
  getAPIBaseURL: () => string,
  pairs: PollAPIPair[]
) => {
  const rates = {
    USD: { BTC: { SIMEX: { latest: 0.00521587214628 } } },
    BTC: {}
  };

  const fallbackRate = 0;

  const arbitraryRateEpoch = 1555452000000; // 17th April 2019 00:00
  const todayTimestamp = new Date().setHours(0, 0, 0, 0);

  const hms = date =>
    new Date(date.getTime() - date.getTimezoneOffset() * 60000)
      .toISOString()
      .split("T")[0];

  const buildDays = from => {
    const outputDays = {};
    const baseRate = baseMockBTCRates[from] || fallbackRate;
    const todayDate = new Date(todayTimestamp);
    const date = new Date();

    for (
      date.setTime(todayTimestamp), date.setFullYear(date.getFullYear() - 3);
      date.getTime() <= todayDate.getTime();
      date.setDate(date.getDate() + 1)
    ) {
      const offset = (date.getTime() - arbitraryRateEpoch) / 864000000 / 180;
      outputDays[todayDate - date ? hms(date) : "latest"] = Math.max(
        baseRate * (1 + offset),
        0
      );
    }
    return Promise.resolve(outputDays);
  };

  await delay(100);

  return pairs.reduce((acc, pair) => {
    if (pair.from === "BTC") return acc;
    switch (pair.exchange) {
      case "bad":
        break;
      case "ugly":
        acc.BTC[pair.from] = { ugly: { latest: 0 } };
        break;
      default:
        acc.BTC[pair.from] = {
          [pair.exchange || "fallback"]: {
            ...buildDays(pair.from)
          }
        };
        break;
    }
    return rates;
  }, rates);
};

/**
 * [good]: Will return deterministic rates for the past 3 years for the pair
 * [ugly]: Will respond, but return 0 as the latest rate.
 * [bad]: Will not return anything for the pair.
 */
export const fetchExchangesForPairImplementation = async () => {
  await delay(100);
  return [
    { id: "good", name: "good", website: "#" },
    { id: "ugly", name: "ugly", website: "#" },
    { id: "bad", name: "bad", website: "#" }
  ];
};

export const tickersByMarketCap = [
  "BTC",
  "ETH",
  "XRP",
  "BCH",
  "LTC",
  "USDT",
  "EOS",
  "BNB",
  "XMR",
  "ADA",
  "XLM",
  "LEO",
  "TRX",
  "HT",
  "DASH",
  "ETC",
  "XTZ",
  "MIOTA",
  "LINK",
  "NEO",
  "MKR",
  "USDC",
  "ONT",
  "CRO",
  "ZEC",
  "DOGE",
  "DCR",
  "PAX",
  "BAT",
  "VET",
  "QTUM",
  "TUSD",
  "BTG",
  "OMG",
  "NANO",
  "BTT",
  "LSK",
  "HOT",
  "REP",
  "WAVES",
  "THETA",
  "ICX",
  "DGB",
  "ZRX",
  "NPXS",
  "KMD",
  "IOST",
  "AOA",
  "DAI",
  "BTM",
  "QNT",
  "ZIL",
  "ENJ",
  "AE",
  "GNT",
  "SNT",
  "SNX",
  "MCO",
  "NEXO",
  "GXC",
  "WIN",
  "XZC",
  "WTC",
  "ELF",
  "XMX",
  "WAX",
  "STRAT",
  "REN",
  "EURS",
  "MANA",
  "FCT",
  "DENT",
  "R",
  "NULS",
  "QASH",
  "LRC",
  "NAS",
  "KNC",
  "PAI",
  "BCV",
  "ZEN",
  "DGTX",
  "MATIC",
  "CHZ",
  "WAN",
  "DEX",
  "FET",
  "DGD",
  "AION",
  "SAN",
  "TOMO",
  "ARK",
  "C20",
  "LOOM",
  "ORBS",
  "CRPT",
  "ODE",
  "ENG",
  "BNT",
  "POWR",
  "BIX",
  "ONE",
  "QKC",
  "ANT",
  "TEL",
  "BRD",
  "VALOR",
  "FSN",
  "STORJ",
  "COCOS",
  "PERL",
  "SEELE",
  "PPT",
  "ABT",
  "CMT",
  "BMC",
  "MTL",
  "TTC",
  "RET",
  "RHOC",
  "CS",
  "CELR",
  "FUN",
  "PIVX",
  "IOTX",
  "EDO",
  "EGT",
  "GRS",
  "MDA",
  "GNO",
  "RLC",
  "AGI",
  "POLY",
  "CVC",
  "EDC",
  "MEDX",
  "VTC",
  "ITC",
  "PART",
  "THR",
  "EKT",
  "NET",
  "CCCX",
  "XET",
  "FXC",
  "QBIT",
  "DTR",
  "BZ",
  "FTT",
  "LA",
  "VERI",
  "BTU",
  "DAPS",
  "CENNZ",
  "SXDT",
  "UTT",
  "TKN",
  "BTR",
  "HPB",
  "CTXC",
  "GNY",
  "EVN",
  "PAY",
  "TNT",
  "PRS",
  "PTT",
  "CBT",
  "KAN",
  "MFT",
  "1ST",
  "RUFF",
  "MAN",
  "HBT",
  "QRL",
  "NOAH",
  "DRGN",
  "PPP",
  "STORM",
  "GTO",
  "HYC",
  "UTK",
  "WABI",
  "UUU",
  "KIN",
  "CPT",
  "SOC",
  "CND",
  "DCN",
  "MWAT",
  "POE",
  "EVX",
  "HYDRO",
  "LBA",
  "SALT",
  "OCEAN",
  "BOX",
  "VITE",
  "COSM",
  "GUSD",
  "GO",
  "INS",
  "DEW",
  "CDT",
  "RDN",
  "PMA",
  "MRPH",
  "PPC",
  "LKY",
  "OST",
  "CHSB",
  "AERGO",
  "REQ",
  "ETHOS",
  "NEC",
  "MIC",
  "MTC",
  "LOCUS",
  "TAAS",
  "QSP",
  "CVT",
  "NCASH",
  "PRO",
  "TEN",
  "RCN",
  "DROP",
  "RNT",
  "LOC",
  "FOAM",
  "APIS",
  "OCN",
  "AUTO",
  "PLR",
  "BLZ",
  "NMR",
  "NPX",
  "MTH",
  "MITX",
  "ADX",
  "MOC",
  "GOT",
  "BZNT",
  "ZIP",
  "WBTC",
  "BET",
  "QUN",
  "JNT",
  "MDT",
  "XSN",
  "OAX",
  "SMT",
  "VEE",
  "VIA",
  "TRV",
  "MET",
  "PLBT",
  "PST",
  "HOT",
  "GVT",
  "CAS",
  "CPT",
  "CHX",
  "IHF",
  "VDG",
  "DMT",
  "DGX",
  "DADI",
  "SRN",
  "NIX",
  "SMART",
  "SWFTC",
  "BOX",
  "BTX",
  "XYO",
  "STACS",
  "SSP",
  "LND",
  "CNN",
  "LYM",
  "KEY",
  "BMX",
  "MLN",
  "OLT",
  "LEND",
  "EDR",
  "RBLX",
  "SWM",
  "DNT",
  "AVA",
  "SENT",
  "RFR",
  "SNGLS",
  "GNX",
  "BCPT",
  "GSC",
  "TIOX",
  "CPC",
  "VIBE",
  "TEMCO",
  "GTC",
  "YOYOW",
  "AST",
  "HMC",
  "DOCK",
  "ECOM",
  "SUSD",
  "UPP",
  "MDS",
  "APPC",
  "WPR",
  "FUEL",
  "BTO",
  "USDS",
  "WINGS",
  "ARN",
  "SLT",
  "SUB",
  "NIM",
  "GET",
  "TRAC",
  "SNC",
  "UBQ",
  "SOUL",
  "SNM",
  "VIB",
  "RPL",
  "INSTAR",
  "VIDT",
  "POA",
  "GEN",
  "RTH",
  "CZR",
  "DLT",
  "SKM",
  "BWX",
  "CREDO",
  "MVL",
  "PCH",
  "CAG",
  "AMO",
  "FTC",
  "ZCL",
  "NEU",
  "ADT",
  "DICE",
  "EOSDAC",
  "ABYSS",
  "ART",
  "XAUR",
  "IDH",
  "YEED",
  "CSNO",
  "UKG",
  "LIF",
  "GENE",
  "EDG",
  "GAME",
  "AMB",
  "GARD",
  "LEO",
  "IHT",
  "VIN",
  "ENQ",
  "UOS",
  "MGO",
  "MXC",
  "BOLT",
  "REM",
  "DNA",
  "YEE",
  "LUN",
  "BKX",
  "CHP",
  "BC",
  "FLIXX",
  "UCASH",
  "TFD",
  "ORS",
  "FTI",
  "XDCE",
  "NCT",
  "AMPL",
  "NOIA",
  "DAX",
  "GRID",
  "AXPR",
  "DTX",
  "MVP",
  "MRK",
  "NOS",
  "WCT",
  "ISR",
  "SPANK",
  "BLT",
  "PRG",
  "IMT",
  "UBT",
  "MTN",
  "CLO",
  "NCC",
  "PBT",
  "OPQ",
  "PCL",
  "CV",
  "LIKE",
  "DAT",
  "NGC",
  "SXUT",
  "BST",
  "ZCN",
  "CXO",
  "EBC",
  "XST",
  "ALIS",
  "1SG",
  "CAN",
  "NBC",
  "KICK",
  "BTT",
  "TRST",
  "BBK",
  "AMLT",
  "PNK",
  "INT",
  "EXRN",
  "PTON",
  "ZCO",
  "RUNE",
  "COV",
  "BAX",
  "EKO",
  "LEDU",
  "FNKOS",
  "KRL",
  "TAU",
  "FXT",
  "HVN",
  "ZLA",
  "ZPR",
  "EDN",
  "MAS",
  "STK",
  "DACS",
  "TFL",
  "CLB",
  "SPD",
  "AID",
  "CPAY",
  "SHIP",
  "QBX",
  "CEEK",
  "HMQ",
  "FTX",
  "CLN",
  "BTCP",
  "FDZ",
  "VSF",
  "CBC",
  "BC",
  "GAT",
  "TIME",
  "PRE",
  "SENSE",
  "INXT",
  "REAL",
  "DREAM",
  "VZT",
  "SPHTX",
  "PARETO",
  "DATX",
  "HKN",
  "UQC",
  "NTK",
  "ROCK2",
  "CARD",
  "ATL",
  "FOTA",
  "MYST",
  "PLU",
  "MTX",
  "CBC",
  "BLUE",
  "ONOT",
  "CAPP",
  "LML",
  "NUG",
  "DBET",
  "BNTY",
  "LCS",
  "TMT",
  "ZSC",
  "IPL",
  "LIFE",
  "RED",
  "BEZ",
  "NANJ",
  "EVN",
  "UP",
  "THRT",
  "DAB",
  "RMESH",
  "X8X",
  "SENC",
  "COFI",
  "DRT",
  "STX",
  "FLP",
  "COIN",
  "TRC",
  "DIT",
  "ETK",
  "BBC",
  "MESG",
  "STQ",
  "PRIX",
  "WYS",
  "BANCA",
  "IETH",
  "NBAI",
  "MNTP",
  "ESS",
  "DXT",
  "PIRL",
  "FTT",
  "PTOY",
  "WIB",
  "UFR",
  "XES",
  "FKX",
  "AVT",
  "BCDT",
  "SWT",
  "BOXX",
  "TCH",
  "SVD",
  "OPEN",
  "RISE",
  "ETBS",
  "PASS",
  "ELEC",
  "VIT",
  "ATX",
  "GENE",
  "RATING",
  "FYP",
  "SS",
  "PHI",
  "1WO",
  "LEV",
  "EBTC",
  "ZMN",
  "CPY",
  "PKT",
  "OMX",
  "MTC",
  "XNK",
  "XPA",
  "PLAY",
  "PAL",
  "SKB",
  "ATM",
  "ZIPT",
  "AID",
  "ESN",
  "EQL",
  "AMN",
  "C8",
  "GOT",
  "GEM",
  "SPN",
  "IVY",
  "SPD",
  "COB",
  "CRC",
  "CL",
  "ZAP",
  "DAV",
  "PYLNT",
  "MAC",
  "TGAME",
  "ERT",
  "EXY",
  "METM",
  "TTV",
  "DRPU",
  "OWN",
  "TNS",
  "FACE",
  "DOV",
  "STAR",
  "ACE",
  "BETHER",
  "OPTI",
  "ZXC",
  "IDXM",
  "XBP",
  "MFG",
  "NIO",
  "ONL",
  "ALI",
  "ADB",
  "BBO",
  "BPT",
  "VLD",
  "BDG",
  "ERO",
  "TBX",
  "OLE",
  "RLX",
  "ETHO",
  "HNST",
  "TGT",
  "PXG",
  "J8T",
  "MOT",
  "RVT",
  "DTH",
  "PIX",
  "JC",
  "BSD",
  "TEAM",
  "HBZ",
  "BOB",
  "HGT",
  "MFTU",
  "HB",
  "TIE",
  "CMCT",
  "PTC",
  "CRED",
  "SND",
  "IXT",
  "UCN",
  "MORE",
  "POLL",
  "HORSE",
  "EXP",
  "ADI",
  "BTR",
  "HER",
  "RCT",
  "JET",
  "SNTR",
  "IND",
  "INV",
  "LATX",
  "VIEW",
  "ONE",
  "ONG",
  "CCRB",
  "CPC",
  "ADH",
  "REBL",
  "CBIX",
  "CAT",
  "FMF",
  "KUE",
  "CAT",
  "EGEM",
  "SNOV",
  "NOX",
  "GUP",
  "BETR",
  "BBN",
  "WAB",
  "LALA",
  "ATMI",
  "TRAK",
  "DEAL",
  "GBX",
  "ARB",
  "WISH",
  "TIX",
  "SMS",
  "PKG",
  "WRC",
  "SET",
  "REF",
  "REX",
  "IFT",
  "SOAR",
  "XMCT",
  "BBK",
  "XPAT",
  "IPSX",
  "BON",
  "MORE",
  "CTX",
  "LUC",
  "MEC",
  "BEE",
  "ADL",
  "EVC",
  "NAVI",
  "SGR",
  "BOUTS",
  "ARY",
  "IG",
  "SIG",
  "LDC",
  "ING",
  "PIPL",
  "ELY",
  "NOBS",
  "DTRC",
  "NTK",
  "GBT",
  "FLOT",
  "SCL",
  "ARCT",
  "EPY",
  "MTRC",
  "OPT",
  "WEB",
  "BTDX",
  "CYFM",
  "FT",
  "FREC",
  "AUC",
  "WEB",
  "ARC",
  "VIU",
  "BERRY",
  "DAN",
  "EXMR",
  "SUR",
  "VRS",
  "ZEUS",
  "ELIX",
  "CDX",
  "XBL",
  "TCH",
  "BTK",
  "MNE",
  "ALV",
  "XCLR",
  "HST",
  "BTCL",
  "VSL",
  "CJT",
  "KNT",
  "AIX",
  "ESZ",
  "CPC",
  "SGN",
  "ORI",
  "DGS",
  "SKIN",
  "EVE",
  "BEAT",
  "BTCONE",
  "ALX",
  "LOCI",
  "JOT",
  "BIT",
  "STAC",
  "KNDC",
  "YUP",
  "SOUL",
  "RLT",
  "PAT",
  "ELLA",
  "STU",
  "SPF",
  "ELTCOIN",
  "EARTH",
  "ZINC",
  "FND",
  "TRCT",
  "OTN",
  "RNTB",
  "ACC",
  "SPX",
  "ITT",
  "PFR",
  "CRB",
  "LNC",
  "HAND",
  "VOISE",
  "GUESS",
  "KWATT",
  "KNT",
  "HODL",
  "CTL",
  "BTRN",
  "CARE",
  "CYMT",
  "TIC",
  "OCC",
  "CMCT",
  "SOL",
  "CCO",
  "RIYA",
  "BNC",
  "DOW",
  "CCT",
  "KIND",
  "AKA",
  "BLN",
  "NDX",
  "FNTB",
  "ATH",
  "ETG",
  "XOV",
  "PMNT",
  "TRDT",
  "FTXT",
  "ARB",
  "VIKKY",
  "GRMD",
  "EVC",
  "SHP",
  "ATS",
  "ICON",
  "CMT",
  "DALC",
  "DELTA",
  "BNN",
  "BIT",
  "STR",
  "BRAT",
  "CLM",
  "HMC",
  "BTX",
  "PLG",
  "ONG",
  "WIN",
  "STC",
  "VD",
  "CET",
  "FNB",
  "MIX",
  "DXR",
  "WHEN",
  "TAN",
  "DXG",
  "HINT",
  "CCC",
  "CBM",
  "IIC",
  "WETH",
  "KEY",
  "GET",
  "EDU",
  "FLETA",
  "CAR",
  "PDATA",
  "TMTG",
  "EMT",
  "HOT",
  "XCHF",
  "GSE",
  "LPT",
  "BCDN",
  "SLT",
  "XTX",
  "ALP",
  "LEMO",
  "ROM",
  "ORS",
  "THAR",
  "CLUB",
  "BIT",
  "BLINK",
  "HLX",
  "CAN",
  "BET",
  "SNIP",
  "EMB",
  "FT",
  "EXC",
  "GZE",
  "DAXT",
  "CCC",
  "TOK",
  "CTT",
  "DXN",
  "ACC"
];

export const fetchTickersByMarketcapImplementation = async (): Promise<
  string[]
> => {
  await delay(100);
  return tickersByMarketCap;
};
