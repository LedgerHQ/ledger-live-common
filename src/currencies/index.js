//@flow
import { encodeURIScheme, decodeURIScheme } from "./CurrencyURIScheme";
import type { Currency } from "../types";

import { sanitizeValueString } from "./sanitizeValueString";

import {
  listFiatCurrencies,
  findFiatCurrencyByTicker,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker
} from "../data/fiat";

import {
  listCryptoCurrencies,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  findCryptoCurrencyById,
  findCryptoCurrencyByScheme,
  findCryptoCurrencyByTicker
} from "../data/cryptocurrencies";

import {
  listTokens,
  listTokensForCryptoCurrency,
  findTokenByTicker,
  findTokenById,
  hasTokenId,
  getTokenById
} from "../data/tokens";

import { parseCurrencyUnit } from "./parseCurrencyUnit";

import { chopCurrencyUnitDecimals } from "./chopCurrencyUnitDecimals";

import {
  formatCurrencyUnit,
  formatCurrencyUnitFragment
} from "./formatCurrencyUnit";

import { formatShort } from "./formatShort";

import { valueFromUnit } from "./valueFromUnit";

const findCurrencyByTicker = (ticker: string): ?Currency =>
  findCryptoCurrencyByTicker(ticker) ||
  findFiatCurrencyByTicker(ticker) ||
  findTokenByTicker(ticker);

export {
  listFiatCurrencies,
  listCryptoCurrencies,
  getFiatCurrencyByTicker,
  findCurrencyByTicker,
  findCryptoCurrencyById,
  findCryptoCurrencyByTicker,
  findCryptoCurrencyByScheme,
  findFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
  listTokens,
  listTokensForCryptoCurrency,
  findTokenByTicker,
  findTokenById,
  hasTokenId,
  getTokenById,
  parseCurrencyUnit,
  chopCurrencyUnitDecimals,
  formatCurrencyUnit,
  formatCurrencyUnitFragment,
  formatShort,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  encodeURIScheme,
  decodeURIScheme,
  valueFromUnit,
  sanitizeValueString
};
