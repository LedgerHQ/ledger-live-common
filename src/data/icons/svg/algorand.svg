<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M16 15.9973H13.4915L11.8624 9.94596L8.35979 15.9973H5.55932L10.973 6.63109L10.1017 3.37855L2.80181 16H0L9.25122 0H11.704L12.778 3.97503H15.3086L13.5808 6.97491L16 15.9973Z" fill="currentColor"/>
</svg>
