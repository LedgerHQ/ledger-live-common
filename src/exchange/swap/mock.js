// @flow

import { BigNumber } from "bignumber.js";
import type {
  Exchange,
  ExchangeRate,
  GetMultipleStatus,
  GetProviders,
  SwapRequestEvent,
} from "./types";
import { getAccountUnit } from "../../account";
import type { Transaction } from "../../types";
import {
  SwapExchangeRateAmountTooLow,
  SwapExchangeRateAmountTooHigh,
} from "../../errors";
import { Observable, of } from "rxjs";

export const mockGetExchangeRates = async (
  exchange: Exchange,
  transaction: Transaction
) => {
  const { fromAccount, toAccount } = exchange;
  const amount = transaction.amount;
  const unitFrom = getAccountUnit(fromAccount);
  const unitTo = getAccountUnit(toAccount);
  const amountFrom = amount.div(BigNumber(10).pow(unitFrom.magnitude));
  const minAmountFrom = BigNumber(0.0001);
  const maxAmountFrom = BigNumber(1000);

  if (amountFrom.lte(minAmountFrom)) {
    throw new SwapExchangeRateAmountTooLow(null, {
      unit: unitFrom.code,
      minAmountFrom,
    });
  }

  if (amountFrom.gte(maxAmountFrom)) {
    throw new SwapExchangeRateAmountTooHigh(null, {
      unit: unitFrom.code,
      maxAmountFrom,
    });
  }

  //Fake delay to show loading UI
  await new Promise((r) => setTimeout(r, 800));

  //Mock OK, not really magnitude aware
  return [
    {
      rate: BigNumber("1"),
      magnitudeAwareRate: BigNumber(1)
        .div(BigNumber(10).pow(unitFrom.magnitude))
        .times(BigNumber(10).pow(unitTo.magnitude)),
      rateId: "mockedRateId",
      provider: "changelly",
      expirationDate: new Date(),
    },
  ];
};

export const mockInitSwap = (
  exchange: Exchange,
  exchangeRate: ExchangeRate,
  transaction: Transaction
): Observable<SwapRequestEvent> => {
  return of({
    type: "init-swap-result",
    initSwapResult: {
      transaction,
      swapId: "mockedSwapId",
    },
  });
};

export const mockGetProviders: GetProviders = async () => {
  //Fake delay to show loading UI
  await new Promise((r) => setTimeout(r, 800));

  return [
    {
      provider: "changelly",
      supportedCurrencies: [
        "bitcoin",
        "litecoin",
        "ethereum",
        "tron",
        "ethereum/erc20/omg",
        "ethereum/erc20/0x_project",
        "ethereum/erc20/augur",
      ],
    },
  ];
};

export const mockGetStatus: GetMultipleStatus = async (statusList) => {
  //Fake delay to show loading UI
  await new Promise((r) => setTimeout(r, 800));
  return statusList.map((s) => ({ ...s, status: "finished" }));
};
