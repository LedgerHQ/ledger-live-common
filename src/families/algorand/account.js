// @flow
import invariant from "invariant";
import { BigNumber } from "bignumber.js";
import type { Account, Operation, Unit } from "../../types";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";

function formatOperationSpecifics(op: Operation, unit: ?Unit): string {
  const { rewards } = op.extra;
  return rewards
    ? " REWARDS : " +
        `${
          unit
            ? formatCurrencyUnit(unit, BigNumber(rewards), {
                showCode: true,
                disableRounding: true,
              }).padEnd(16)
            : rewards
        }`
    : "";
}

function formatAccountSpecifics(account: Account): string {
  const { algorandResources } = account;
  invariant(algorandResources, "algorand account expected");
  const unit = getAccountUnit(account);
  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };

  let str = " ";

  str +=
    formatCurrencyUnit(unit, account.spendableBalance, formatConfig) +
    " spendable. ";
  if (algorandResources.rewards.gt(0)) {
    str +=
      formatCurrencyUnit(unit, algorandResources.rewards, formatConfig) +
      " rewards. ";
  }
  if (algorandResources.rewardsAccumulated.gt(0)) {
    str +=
      formatCurrencyUnit(
        unit,
        algorandResources.rewardsAccumulated,
        formatConfig
      ) + " rewardsAccumulated. ";
  }

  return str;
}

export function fromOperationExtraRaw(extra: ?Object) {
  if (extra && extra.rewards) {
    return {
      ...extra,
      rewards: BigNumber(extra.rewards),
    };
  }
  return extra;
}

export function toOperationExtraRaw(extra: ?Object) {
  if (extra && extra.rewards) {
    return {
      ...extra,
      rewards: extra.rewards.toString(),
    };
  }
  return extra;
}

export default {
  formatAccountSpecifics,
  formatOperationSpecifics,
  fromOperationExtraRaw,
  toOperationExtraRaw,
};
