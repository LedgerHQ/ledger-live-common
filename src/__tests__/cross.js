// @flow
import "./test-helpers/staticTime";
import "../load/tokens/ethereum/erc20";
import "../load/tokens/tron/trc10";
import "../load/tokens/tron/trc20";
import "../load/tokens/algorand/asa";

import { genAccount } from "../mock/account";
import { getDerivationModesForCurrency } from "../derivation";
import { listCryptoCurrencies } from "../currencies";
import {
  accountDataToAccount,
  accountToAccountData,
  encode,
  decode,
} from "../cross";

test("accountDataToAccount / accountToAccountData", () => {
  listCryptoCurrencies().forEach((currency) => {
    getDerivationModesForCurrency(currency).forEach((derivationMode) => {
      const account = genAccount(`${currency.id}_${derivationMode}`, {
        currency,
      });
      const data = accountToAccountData(account);
      expect(accountToAccountData(accountDataToAccount(data))).toMatchObject(
        data
      );
    });
  });
});

test("encode/decode", () => {
  const accounts = listCryptoCurrencies().reduce(
    (acc, currency) =>
      acc.concat(
        getDerivationModesForCurrency(currency).map((derivationMode) => {
          const account = genAccount(`${currency.id}_${derivationMode}`, {
            currency,
          });
          return account;
        })
      ),
    []
  );

  const data = {
    accounts,
    settings: {
      currenciesSettings: {},
      pairExchanges: {},
    },
    exporterName: "testä½ å¥½ðŸ‘‹",
    exporterVersion: "0.0.0",
  };
  const exp = decode(encode(data));
  expect(exp.meta.exporterName).toEqual(data.exporterName);
  expect(exp.accounts.length).toEqual(data.accounts.length);
  expect(exp.accounts).toMatchObject(data.accounts.map(accountToAccountData));
});

test("encode/decode swapHistory", () => {
  const accountWithHistory = genAccount("account_with_swapHistory", {
    swapHistorySize: 2,
  });
  const data = accountToAccountData(accountWithHistory);
  const res = accountToAccountData(accountDataToAccount(data));
  expect(res).toMatchObject(data);
  expect(res.swapHistory?.length).toBe(2);
});

test("encode/decode swapHistory resilience", () => {
  // swapHistory should be added even if the source doesn't have it
  const accountWithoutHistory = genAccount("account_without_swapHistory");
  // $FlowFixMe
  accountWithoutHistory.swapHistory = undefined;
  const data = accountToAccountData(accountWithoutHistory);
  const res = accountToAccountData(accountDataToAccount(data));

  expect(res).toMatchObject(data);
  expect(res.swapHistory?.length).toBe(0);
});

test("encode/decode", () => {
  const accounts = Array(3)
    .fill(null)
    .map((_, i) => genAccount("export_" + i));
  const arg = {
    accounts,
    settings: {
      counterValue: "USD",
      pairExchanges: {
        BTC_USD: "KRAKEN",
      },
      currenciesSettings: {
        bitcoin: {
          confirmationsNb: 3,
        },
      },
      blacklistedTokenIds: ["tokenid1", "tokenid2"],
    },
    exporterName: "test",
    exporterVersion: "0.0.0",
    chunkSize: 100,
  };
  const data = encode(arg);
  const res = decode(data);
  expect(res.accounts).toMatchObject(
    accounts.map((a) => ({
      balance: a.balance.toString(),
      currencyId: a.currency.id,
      id: a.id,
      name: a.name,
      index: a.index,
    }))
  );
  expect(res.settings).toMatchObject({
    counterValue: "USD",
    pairExchanges: {
      BTC_USD: "KRAKEN",
    },
    currenciesSettings: {
      bitcoin: {
        confirmationsNb: 3,
      },
    },
    blacklistedTokenIds: ["tokenid1", "tokenid2"],
  });
  expect(res.settings).not.toMatchObject({
    counterValue: "USD",
    pairExchanges: {
      BTC_USD: "KRAKEN",
    },
    currenciesSettings: {
      bitcoin: {
        confirmationsNb: 3,
      },
    },
    blacklistedTokenIds: ["tokenid3"],
  });
  expect(res).toMatchSnapshot();
});
