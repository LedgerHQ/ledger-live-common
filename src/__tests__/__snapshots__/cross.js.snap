// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode/decode 1`] = `
Object {
  "accounts": Array [
    Object {
      "balance": "12121280662072834414",
      "currencyId": "ethereum",
      "derivationMode": "",
      "freshAddress": "0x8C5FE6AF52E79EC4361B7A12B53356EF5D37C765",
      "id": "mock:1:ethereum:export_0:",
      "index": 1,
      "name": "qfLJHUiyMqUUN",
      "seedIdentifier": "mock",
    },
    Object {
      "balance": "53736258",
      "currencyId": "bitcoin_cash",
      "derivationMode": "",
      "freshAddress": "1x7aLVWzaGVH8AWQ4aaD2hdcmJW",
      "id": "mock:1:bitcoin_cash:export_1:",
      "index": 1,
      "name": "Y4SYkWgSWIIZptphvAu",
      "seedIdentifier": "mock",
    },
    Object {
      "balance": "128088571428550",
      "currencyId": "dogecoin",
      "derivationMode": "",
      "freshAddress": "1CWNKLfcukxxXRNRK1DpYwPA6Hb1h",
      "id": "mock:0:dogecoin:export_2:",
      "index": 1,
      "name": "s9dZEiOrwfEPOXbKgNKq88aQZ",
      "seedIdentifier": "mock",
    },
  ],
  "meta": Object {
    "exporterName": "test",
    "exporterVersion": "0.0.0",
  },
  "settings": Object {
    "blacklistedTokenIds": Array [
      "tokenid1",
      "tokenid2",
    ],
    "counterValue": "USD",
    "currenciesSettings": Object {
      "bitcoin": Object {
        "confirmationsNb": 3,
      },
    },
    "pairExchanges": Object {
      "BTC_USD": "KRAKEN",
    },
  },
}
`;
